src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,542 @@
+            return (result == size ? new proc_taskallinfo?(info) : null);
+        }
+
+        /// <summary>
+        /// Gets the thread information for the given thread
+        /// </summary>
+        /// <param name="thread">The ID of the thread to query for information</param>
+        /// <returns>
+        /// Returns a valid proc_threadinfo struct for valid threads that the caller
+        /// has permissions to access; otherwise, returns null
+        /// </returns>
+        internal static unsafe proc_threadinfo? GetThreadInfoById(int pid, ulong thread)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Negative TIDs are invalid
+            if (thread < 0)
+            {
+                throw new ArgumentOutOfRangeException("thread");
+            }
+
+            // Get the thread information for the specified thread in the specified process
+            int size = Marshal.SizeOf<proc_threadinfo>();
+            proc_threadinfo info = default(proc_threadinfo);
+            int result = proc_pidinfo(pid, PROC_PIDTHREADINFO, (ulong)thread, &info, size);
+            return (result == size ? new proc_threadinfo?(info) : null);
+        }
+
+        internal static unsafe List<KeyValuePair<ulong, proc_threadinfo?>> GetAllThreadsInProcess(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            int result = 0;
+            int size = 20; // start assuming 20 threads is enough
+            ulong[] threadIds = null;
+
+            // We have no way of knowning how many threads the process has (and therefore how big our buffer should be)
+            // so while the return value of the function is the same as our buffer size (meaning it completely filled
+            // our buffer), double our buffer size and try again. This ensures that we don't miss any threads
+            do
+            {
+                threadIds = new ulong[size];
+                fixed (ulong* pBuffer = threadIds)
+                {
+                    result = proc_pidinfo(pid, PROC_PIDLISTTHREADS, 0, pBuffer, Marshal.SizeOf<ulong>() * threadIds.Length);
+                }
+
+                if (result <= 0)
+                {
+                    throw new System.ComponentModel.Win32Exception(Marshal.GetLastWin32Error());
+                }
+                else
+                {
+                    size *= 2;
+                }
+            }
+            while (result == Marshal.SizeOf<ulong>() * threadIds.Length);
+
+            Debug.Assert((result % Marshal.SizeOf<ulong>()) == 0);
+
+            // Remove any extra elements
+            Array.Resize<ulong>(ref threadIds, (int)(result / Marshal.SizeOf<ulong>()));
```

Comment:
Seems like this resize isn't necessary.  You know how many elements there are, and you're about to copy those elements into a list.  So you can just iterate from 0 to the number of elements rather than foreach'ing.  You can also pass that length to the List's ctor to preset the capacity.